File: 0496-next-greater-element-i/0496-next-greater-element-i.cpp
class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        vector<int> result;
        unordered_map<int,int> map;
        stack<int> stack;
        for(int i=0; i<nums2.size();i++){
            while(!stack.empty() && nums2[i]>stack.top()){
                map[stack.top()]=nums2[i];
                    stack.pop();
            }
            
            stack.push(nums2[i]);
        }

        
            for(int i:nums1){
                if(map.count(i)){
                    result.push_back(map[i]);
                }else{
                    result.push_back(-1);
                }
            }
            return result;


    }
};

File: 0496-next-greater-element-i/0496-next-greater-element-i.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
    #     1-0
    #     3-1
    #     4-2
    #     2-3
        
    # 1 3 4 2
    #     4  -1
    # 1 5 2 3 7
    # 0 0 0 7 0
        prefix = [0] * (len(nums2) + 1)
        prefix[-1] = -1
        for i in range(len(nums2)-1,0,-1):
            if nums2[i] > nums2[i-1]:
                prefix[i] = nums2[i]
            else:
                prefix[i] = prefix[i+1]
        count = {}
        for i,n in enumerate(nums2):
            count[n] = i
        ans = []
        for n in nums1:
            idx = count[n]
            ans.append(prefix[idx+1])

        return ans
